esx_service/cli/vmdkops_admin.py

```diff
@@ -726,38 +731,382 @@ def get_version():
+    
+    return privileges
 
 def tenant_access_add(args):
-    print(NOT_IMPLEMENTED)
+    """ Handle tenant access command """
+    error_info, tenant = get_tenant_from_db(args.name)
+    if error_info:
+        return operation_fail(error_info)
+
+    if not tenant:
+        error_info = "Tenant {0} does not exist".format(args.name)
+        return operation_fail(error_info)
+
+    privileges = generate_privileges(args)
+    print privileges
+    error_info = tenant.set_datastore_access_privileges(auth._auth_mgr.conn, [privileges])
+      
+    if error_info:
+        return operation_fail(error_info)
+    else:
+        print "tenant access add succeeded"
+
+def modify_privileges(privileges, args):
+    """ Modify privileges based on CLI argument """
+    if args.add_rights:
+        set_privileges(args.add_rights, privileges, 1)
+        
+    if args.rm_rights:
+        set_privileges(args.rm_rights, privileges, 0)
+
+    if args.volume_maxsize:
+        size_in_MB = convert.convert_to_MB(args.volume_maxsize)
+        privileges[auth_data_const.COL_MAX_VOLUME_SIZE] = size_in_MB
+    
+    if args.volume_totalsize:
+        size_in_MB = convert.convert_to_MB(args.volume_totalsize)
+        privileges[auth_data_const.COL_USAGE_QUOTA] = size_in_MB
+    
+    return privileges
+
+def generate_privileges_dict(privileges):
+    # privileges is a list which is read from auth DB
+    # it has the following format
+    # (tenant_uuid, datastore, global_visibility, create_volume, delete_volume,
+    # mount_volume, max_volume_size, usage_quota)
+    privileges_dict = {}
+    privileges_dict[auth_data_const.COL_DATASTORE] = privileges[1]
+    privileges_dict[auth_data_const.COL_GLOBAL_VISIBILITY] = privileges[2]
+    privileges_dict[auth_data_const.COL_CREATE_VOLUME] = privileges[3]
+    privileges_dict[auth_data_const.COL_DELETE_VOLUME] = privileges[4]
+    privileges_dict[auth_data_const.COL_MOUNT_VOLUME] = privileges[5]
+    privileges_dict[auth_data_const.COL_MAX_VOLUME_SIZE] = privileges[6]
+    privileges_dict[auth_data_const.COL_USAGE_QUOTA] = privileges[7]
+    return privileges_dict
 
 def tenant_access_set(args):
-    print(NOT_IMPLEMENTED)
-
+    """ Handle tenant access set command """
+    error_info, tenant = get_tenant_from_db(args.name)
+    if error_info:
+        return operation_fail(error_info)
+
+    if not tenant:
+        error_info = "Tenant {0} does not exist".format(args.name)
+        return operation_fail(error_info)
+
+    privileges = [d for d in tenant.privileges if d[auth_data_const.COL_DATASTORE] == args.datastore]
+    
+    if not privileges:
+        error_info = "No privileges exist for ({0}, {1})".format(args.name, args.datastore)
+        return operation_fail(error_info)
+    
+    privileges_dict = generate_privileges_dict(privileges[0])
+    privileges_dict = modify_privileges(privileges_dict, args)
+    
+    error_info = tenant.set_datastore_access_privileges(auth._auth_mgr.conn, [privileges_dict])
```

Comment:
I assume Read-After-Write allows user to selectively modify subset of permissions and doesn't overwrite existing privileges.